### ISSUE TITLE
Ancestral state reconstruction infers ambiguous tip states by default, contrary to documentation

### ISSUE THREAD

**Issue by jameshadfield** [nextstrain/augur#612]
The current documentation for `augur ancestral` (augur v10) makes it clear that ambiguous sites are not reconstructed by default:

```
  --keep-ambiguous      do not infer nucleotides at ambiguous (N) sites on tip
                        sequences (leave as N). (default: True)
  --infer-ambiguous     infer nucleotides at ambiguous (N,W,R,..) sites on tip
                        sequences and replace with most likely state.
                        (default: False)
```

However this is not the case. Internally, the salient variable is `args.infer_ambiguous` and the current behavior is:

| CLI Argument (s)        | `args.infer_ambiguous`         | Behavior Expected?  |
| ------------- |-------------| -----|
| `--infer-ambiguous`     | True | Yes |
| `--keep-ambiguous` | False      |   Yes |
| `--infer-ambiguous --keep-ambiguous` | _Fatal Error_      |    Yes |
| (no argument) | True     |    No |

We can fix this by either changing the set-up of arguments or change the documentation to say that tips _are_ reconstructed by default. cc @huddlej which approach would you recommend?

**Comment by emmahodcroft** [comment]
Tagging @rneher as I know this is something we discussed a lot, though I don't really remember which way we settled.

**Comment by rneher** [comment]
the code doesn't specify an explicit default. We decided we wanted to infer ambiguous sites by default to not change previous behavior. Otherwise I would opt for not reconstructing them by default.

**Comment by huddlej** [comment]
Since this is an issue of how the defaults are documented rather than which defaults we use, I'd vote for not changing the default behavior to address this issue.

I'd recommend changing the argument setup such that the help documentation prints the correct information and the code is clearer to read for us as maintainers. Specifically, I'd make the changes in the following patch such that:

 - the two ambiguity flags store their values in separate variables (avoiding conflicting defaults and making the values of the variables easier to determine by reading the code)
 - the `--keep-ambiguous` flag has an implicit default of `False` because of its action type (the usual behavior for this type of opt-in flag)
 - the `--infer-ambiguous` flag has an explicit default of `True` to properly document the default behavior in the help output and communicate to us as developers that this flag is essentially redundant
 - the script determines whether to infer ambiguity based on the joint logic of the provided flags

```diff
diff --git a/augur/ancestral.py b/augur/ancestral.py
index 947aecc..805eb06 100644
--- a/augur/ancestral.py
+++ b/augur/ancestral.py
@@ -111,9 +111,9 @@ def register_arguments(parser):
     parser.add_argument('--vcf-reference', type=str, help='fasta file of the sequence the VCF was mapped to')
     parser.add_argument('--output-vcf', type=str, help='name of output VCF file which will include ancestral seqs')
     ambiguous = parser.add_mutually_exclusive_group()
-    ambiguous.add_argument('--keep-ambiguous', action="store_false", dest='infer_ambiguous',
+    ambiguous.add_argument('--keep-ambiguous', action="store_true",
                                 help='do not infer nucleotides at ambiguous (N) sites on tip sequences (leave as N).')
-    ambiguous.add_argument('--infer-ambiguous', action="store_true",
+    ambiguous.add_argument('--infer-ambiguous', action="store_true", default=True,
                                 help='infer nucleotides at ambiguous (N,W,R,..) sites on tip sequences and replace with most likely state.')
     parser.add_argument('--keep-overhangs', action="store_true", default=False,
                                 help='do not infer nucleotides for gaps (-) on either side of the alignment')
@@ -159,9 +159,14 @@ def run(args):
         print("ERROR: this version of augur requires TreeTime 0.7 or later.")
         return 1

+    # Infer ambiguous bases if the user has requested that we infer them (either
+    # explicitly or by default) and the user has not explicitly requested that
+    # we keep them.
+    infer_ambiguous = args.infer_ambiguous and not args.keep_ambiguous
+
     tt = ancestral_sequence_inference(tree=T, aln=aln, ref=ref, marginal=args.inference,
                                       fill_overhangs = not(args.keep_overhangs),
-                                      infer_tips = args.infer_ambiguous)
+                                      infer_tips = infer_ambiguous)

     character_map = {}
     for x in tt.gtr.profile_map:
@@ -173,7 +178,7 @@ def run(args):
             character_map[x] = x

     anc_seqs['nodes'] = collect_mutations_and_sequences(tt, full_sequences=not is_vcf,
-                            infer_tips=args.infer_ambiguous, character_map=character_map)
+                            infer_tips=infer_ambiguous, character_map=character_map)
     # add reference sequence to json structure. This is the sequence with
     # respect to which mutations on the tree are defined.
     if is_vcf:
```

We should be able to include this fix in a patch release 10.0.1, if there aren't any objections to the above changes.

**Comment by jameshadfield** [comment]
Thanks @huddlej -- this seems good to me. From the user's perspective, the only change is to the following help documentation which now matches the (unchanged) behavior:

```
  --keep-ambiguous      do not infer nucleotides at ambiguous (N) sites on tip
                        sequences (leave as N). (default: False)
  --infer-ambiguous     infer nucleotides at ambiguous (N,W,R,..) sites on tip
                        sequences and replace with most likely state.
                        (default: True)
```
